lib/cdo/redshift_import.rb

```diff
@@ -0,0 +1,170 @@
+  # "_import_" to avoid disrupting usage of the Analytics/Reporting solutions during the lengthy Full Load process.
+  # Complete import of staged data by renaming the current table to "_old_", renaming the staging table, and then
+  # dropping the old table.
+  # @param schemas [String] Array of Redshift schemas in which to complete data import.
+  def self.complete_table_import(schemas)
+    schemas.each do |schema|
+      temporary_import_tables(schema).each do |import_table|
+        target_table = import_table.partition(TEMP_TABLE_PREFIX).last
+        backup_table = BACKUP_TABLE_PREFIX + target_table
+
+        CDO.log.info "Dropping existing table #{schema}.#{target_table} and renaming newly imported #{import_table}."
+
+        # Rename existing table to back it up, if it exists.
+        begin
+          rename_table(schema, target_table, backup_table)
+        rescue PG::UndefinedTable => undefined_table_error
+          # When a new table created in the source MySQL database is imported for the first time, there won't be an
+          # existing table in Redshift to backup.
+          CDO.log.info "Unable to back up table #{schema}.#{target_table} because it does not exist. #{undefined_table_error}"
+        end
+
+        # Make staging table the production table.
+        rename_table(schema, import_table, target_table)
+
+        # Rename the primary key from '_import_[foo]_primary' to '[foo]_primary' so that the next time the
+        # DMS Replication Task runs and creates the staging table it can create the primary key without being blocked
+        # by an existing primary key with that name.
+        # Also rename the backup table's primary key.
+        old_primary_key = RedshiftImport.primary_key(schema, backup_table)
+        if old_primary_key
+          rename_primary_key(
+            schema,
+            backup_table,
+            old_primary_key[:name],
+            BACKUP_TABLE_PREFIX + old_primary_key[:name],
+            old_primary_key[:columns]
+          )
+        end
+
+        primary_key = RedshiftImport.primary_key(schema, target_table)
+        if primary_key
+          rename_primary_key(
+            schema,
+            target_table,
+            primary_key[:name],
+            primary_key[:name].partition(TEMP_TABLE_PREFIX).last,
+            primary_key[:columns]
+          )
+        end
+
+        # Drop the old production table.
+        drop_table(schema, backup_table)
+      end
+    end
+  end
+
+  # Get names of tables in a Redshift schema that were loaded via DMS to stage imported data from the database.
+  # @param schema [String] The Redshift schema to search for database import staging tables.
+  # Returns Array of Redshift table names.
+  def self.temporary_import_tables(schema)
+    redshift_client = RedshiftClient.instance
+    query = <<~SQL
+      SET search_path TO #{schema};
```

Comment:
Do we need to set `search_path` here for querying `pg_table_def` and `pg_indexes`?

Reply:
I thought this was something I accidentally left here while copying/pasting from other queries, but it turns out that it is required to ensure the correct schema is queried even though we're supposedly directly inspecting the Postgres schema tables.